merkle/index.go

```diff
@@ -2,52 +2,83 @@ package merkle
 
 // siblingIndecies returns indecies of sibling elements
 func siblingIndecies(idxs []uint64) []uint64 {
+
 	indicesCount := len(idxs)
 	siblings := make([]uint64, indicesCount)
+
+	// append all sibling indices to result
 	for i := 0; i < indicesCount; i++ {
 		siblings[i] = siblingIndex(idxs[i])
 	}
+
 	return siblings
 }
 
 // parentIndecies returns indecies of parent elements
 func parentIndecies(idxs []uint64) []uint64 {
-	indicesCount := len(idxs)
+
 	var parents []uint64
-	for i := 0; i < indicesCount; i++ {
+
+	// loop through all indices
+	for i := 0; i < len(idxs); i++ {
+
+		// get parent index
 		parentIndex := parentIndex(idxs[i])
+
+		// checl if it is duplicate
 		isDuplicate := false
 		for j := 0; j < len(parents); j++ {
```

Comment:
**Maybe** you don't need this loop at all and hence can run in O(n).
My intuition is the following:
if you have a set of indices, and want to get their parents without any duplication you can:
- get all the even indices on the level
- for the odd indices, check that you don't have a sibling that's equal to itself - 1.

Reply:
@vjdmhd 

```diff
 func parentIndecies(idxs []uint64) []uint64 {
        indicesCount := len(idxs)
        var parents []uint64
+       indexFound := make(map[uint64]struct{})
        for i := 0; i < indicesCount; i++ {
                parentIndex := parentIndex(idxs[i])
-               isDuplicate := false
-               for j := 0; j < len(parents); j++ {
-                       if parentIndex == parents[j] {
-                               isDuplicate = true
-                       }
-               }
+               _, isDuplicate := indexFound[parentIndex]
                if !isDuplicate {
                        parents = append(parents, parentIndex)
+                       indexFound[parentIndex] = struct{}{}
                }
-
        }
        return parents
 }
```

However, it doesn't really improve `BenchmarkParentIndices-24`. Please double check,

Reply:
Check my suggestion here: https://github.com/ComposableFi/go-merkle-trees/pull/25